azkaban-exec-server/src/main/java/azkaban/execapp/ProjectCacheCleaner.java

```diff
@@ -40,16 +43,18 @@
 class ProjectCacheCleaner {
```

Comment:
I think I already mentioned in an earlier PR, but longer term it would be good to move cache cleanup to a different thread. This can be expensive and time-consuming, and isn't necessary as part of flow preparation itself. Moving cleanup out of flow preparation would help speed up dispatch.

One option would be to have a ConcurrentHashMap with key as <project id, version> and value as a Lock. The lock would then be at the project/version level, so flows accessing different projects would not need to block on each other. The lock would be gotten as part of flow preparation and cleanup for the project.

The lock could also be done as a ReentrantReadWriteLock, but if the time in the synchronized section is small (would usually just be the time to hardlink), then this may not add a lot of value. In this case flow preparation would get a read lock (creating the lock if it does not exist) and check if the project directory already exists. If the directory does not exist, then upgrade to a write lock, check again for directory existence,  create the project directory if needed, then downgrade to write lock. Cleanup would get the write lock, and delete the directory. It would be good for cleanup to remove the lock from the hashmap as well, but this would be trickier with concurrent readers, so we'd want to do some testing.

Note that this would only handle concurrency within the same JVM. For upgrade, when there may be 2 executors (old and new), then we would need to be careful that they are not both modifying the cache (creating new project directories or removing them) at the same time, such as your PR to wait during executor invalidation. The old executor should also be prevented from cleaning up the cache. 

Another option could be to use directories and files within the project directory. There could be a .lock directory, and each thread could create a file with its <executor id, thread id>. Cleanup would remove the empty .lock directory.

Reply:
yes, i agree it would be best to have project id/version as the locking key and move the clean up to another thread.  Let me think about how to address this problem.

Reply:
I'm planning to move the flow preparation logic to the execution thread, and will need some of the same data structures, so can take this up.